pkg/cache/cluster.go

```diff
@@ -568,30 +571,31 @@ func (c *clusterCache) listResources(ctx context.Context, resClient dynamic.Reso
 }
 
 func (c *clusterCache) loadInitialState(ctx context.Context, api kube.APIResourceInfo, resClient dynamic.ResourceInterface, ns string, lock bool) (string, error) {
-	return c.listResources(ctx, resClient, func(listPager *pager.ListPager) error {
-		var items []*Resource
-		err := listPager.EachListItem(ctx, metav1.ListOptions{}, func(obj runtime.Object) error {
+	var items []*Resource
```

Comment:
By moving this outside the semaphore, we increase the simultaneous memory usage, right? Will this increase the memory spike on cache construction?

Reply:
The only other way I can think of processing the chunk of items with a limit of 50x api items in memory is to add the exception like @jdroot proposed in https://github.com/argoproj/gitops-engine/pull/599. If you hold the lock already, you can bypass the 50x semaphore, and because startMissingWatches is not processing all watches in parallel like sync, it is not that impactful.

I just added a unit test that can reproduce the deadlock and also a few others that were already fixed in previous PRs. Maybe it will help find out if startMissingWatches needs to hold the lock for so long.

Reply:
Okay I think I'm in favor of merging to fix the deadlock and then deal with any memory consumption regression if it appears.

Reply:
> Maybe it will help find out if startMissingWatches needs to hold the lock for so long, which would also solve the issue elegantly.

This would be the real ideal fix I think. Minimizing lock holding time is generally a good practice. I see what you mean about the memory issue this proposed change causes, I wonder how much that will work out to be in practice, and how spikey it will be.

Reply:
imo since the semaphore is per-cluster, we're probably not mitigating spikes very well anyway. Personal experience from monitoring memory use in rollouts backs up that feeling.

Reply:
We deployed v2.11.5 last night, which I believe includes this commit. Our clusters are quite large (some over 1000 nodes, with 1000s of namespaces, 10s of thousands of pods) and noticed no meaningful spike in memory (either average or during reconciliation). So I think the fix looks pretty solid from that perspective. It was difficult to reproduce the problem, so I am not sure I have a good way to validate the deadlock no longer happens

Reply:
Nice. The main thing to look for is whether the workqueue depth ever gets stuck with a "floor," i.e. the deadlocked cluster's queue depth.

Reply:
We have monitors and pages set up for if we hit issues, but lack of an issue isn't always indicative of the problem being fixed haha